Unified and Customizable Finetuning callback using hooks and registry. (#830)
PyTorchLightning/lightning-flash: docs/source/general/finetuning.rst

freeze_unfreeze
---------------

> The freeze_unfreeze strategy keeps the backbone frozen until a certain epoch (provided through the input) after which the backbone unfrozen.

Reviewer: Maybe something like:
```suggestion
The ``freeze_unfreeze`` strategy keeps the backbone frozen until a certain epoch (provided in a tuple to the ``strategy`` argument) after which the backbone will be unfrozen.
```

Custom Strategy
===============
Flash provides customizability for finetuning in two ways.

custom
------
This strategy allows you to implement a custom finetuning logic using the freezing logic provided by Flash out of the box.
The previosly mentioned strategies either unfreezes the complete backbone or just a fixed fraction of the backbone.
This stratgey provides you much more control on that logic during finetuning.

Here's an example where:

> * a distilbert for question answering backbone starts frozen

Reviewer: ```suggestion
* a distilbert for question answering backbone starts frozen (the default behaviour)
```

> * at epoch 2 just the transformer unfreezes

Reviewer: ```suggestion
* we override the ``finetune_backbone`` hook so that at epoch 2 just the transformer unfreezes
```

|

.. code-block:: python

    from flash import Trainer
    from flash.core.data.utils import download_data
    from flash.text import QuestionAnsweringData, QuestionAnsweringTask
    from torch.optim import Optimizer

    download_data("https://pl-flash-data.s3.amazonaws.com/squad_tiny.zip", "./data/")

    datamodule = QuestionAnsweringData.from_squad_v2(
        train_file="./data/squad_tiny/train.json",
        val_file="./data/squad_tiny/val.json",
    )

    class MyCustomQuestionAnsweringTask(QuestionAnsweringTask):

>         def __init__(self, **kwargs):
>             super().__init__(**kwargs)

Reviewer: This can be removed:
```suggestion
```

>         def finetune_backbone(self, epoch: int, optimizer: Optimizer, opt_idx: int) -> None:

Reviewer: I think it might be cleaner for this hook to stay on the finetuning callback. That way this class which implements only custom finetuning logic would not be a new task, just a different strategy to use with the same task (and then the callback should be passed to the strategy argument).

Author: The idea behind `finetune_backbone` is that if we implement it in the `finetune_function` of the `FlashBaseFinetuning` class by creating a new custom callback, we don't get to use the `freeze_before_training` logic provided by Flash.

Reviewer: What about if we had the current finetune_backbone hook just as part of the finetune callback rather than the model? That way the finetune callbacks could be used outside of flash and just implement the finetune_backbone hook to add custom unfreezing behaviour?

Reviewer: Also, should the hook maybe be called something like unfreeze_at_epoch(epoch: int, ...) to make the intended usage (unfreezing some weights) more clear?

Author: We can make the custom mode a callable based on.

```
def my_unfreezing_logic(pl_module, epoch, optimizer, opt_idx):
    <perform the logic>
trainer.finetune(model, datamodule, strategy=my_unfreezing_logic)
```

So then we don't need the `custom` key and directly infer that it's a custom function using existing freezing logic.

Reviewer 2: I would prefer not to have this on the task personally. Not all models have a backbone. Or we should make it more general. `on_finetune_start`

>             self.model.distilbert.transformer.requires_grad_(False)

Reviewer: Should this be:
```suggestion
            if epoch == 2:
                self.model.distilbert.transformer.requires_grad_(False)
```
